{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**<center style=\"color:#5F018B;font-size:24pt;weight:bold\">SCIE2100 Practical 0 (Weeks 1-2):</center>**\n",
    "\n",
    "**<center style=\"color:#5F018B;font-size:20pt;weight:bold\">Computer programming</center>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Aim</SPAN>**\n",
    "\n",
    "**After reading and completing this practical, you should:**\n",
    "* Know the basics of the Python programming language\n",
    "* Know how to run Python programs\n",
    "* Know how to edit Python programs\n",
    "* Be familiar with how to access data in regards to both file types and web resources\n",
    "\n",
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Outline</SPAN>**\n",
    "\n",
    "**Introduction**\n",
    "* Semester Overview\n",
    "* The Python Guide\n",
    "* Accessing the practicals\n",
    "* Practical 0\n",
    "\n",
    "**Part 1 : Reviewing Python basics**\n",
    "* Programming lingo - let's start from the bottom and work our way up\n",
    "* Exercises 1-4\n",
    "\n",
    "**Part 2 : More complex Python structures**\n",
    "* Working with Classes\n",
    "* Exercises 5-14\n",
    "\n",
    "**Part 3: Accessing online files and databases**\n",
    "* Automating access to online resources\n",
    "* FASTA file format\n",
    "* Exercises 15-18"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Introduction</SPAN>**\n",
    "\n",
    "### Semester Overview\n",
    "\n",
    "Over the course of this semester, you will complete practicals that will not only reinforce the material from lectures, but also give you experience in performing bioinformatics tasks. **You must attend the practical each week**, during which you will follow a guide like this one that you can download from the Blackboard site.  While completing these practicals, you will often write Python code to analyse biological data, use web resources, and learn about common file formats for data storage. You will also receive guidance from tutors to help you in achieveing the learning goals for each week's practical this semester. Make sure you ask the tutors questions when needed and that you take advantage of this resource.\n",
    "\n",
    "In total, you will complete seven practicals (Practicals 0-6) and three practical assessments (Practical Assessment 1-3). Practical 0 - which you will be working on for the first two weeks - *will not be assessed for a grade* and will solely serve as a practice and introduction to the practical component of this course. The six subsequent practicals (Practicals 1-6) will each be worth 3% of your grade (18% in total) and the three practical assessments will each be worth 10% of your grade (30% in total). You will be graded on your ability to work through the exercises in the practical and on your answers to questions about what you have done, demonstrating analytic thinking about the principles involved (see the course profile for details of the official assessment criteria).\n",
    "\n",
    "### The Python Guide\n",
    "\n",
    "We have provided a Python Guide that is specific to this course as a reference tool for you to use throughout the semester. If you forget how to write a for loop or forget what a dictionary is, for example, you can use it to quickly remind yourself. Each section of the guide also contains references to the corresponding part of the official Python online tutorial in case you want to read more, or even if you just want the concept explained in a different way. Also don’t forget that Python is a very popular programming language, and that if you search the web for any problems you have, you’ll probably find an answer.\n",
    "\n",
    "Read through the reference guide now to familiarise yourself with Python we’ll be using in this course, and to reacquaint yourself with those bits of Python that you already know. As you go, type out and run the code examples yourself to make sure you understand them. If you have any problems or questions, ask the tutor. It is better to sort out any problems you have now rather than later.\n",
    "\n",
    "### Accessing the practicals\n",
    "\n",
    "All the files you need for each week's practical will be available for download from the course's Blackboard site. Each practical sheet, like this one, will list all the files you need and explain what they contain. For this practical, you need this file, the Python Guide and a couple of Python files (introduction.py, bonus_code.py, guide.py and translate.py). The introduction and bonus_code Python files cover very similar topics but with slightly different examples.\n",
    "\n",
    "### Practical 0\n",
    "\n",
    "This practical will serve as an introduction to the course, the computers in the lab, and the Python programming language. Since this practical is designed to be completed over the first two weeks of the semester, it is double the size of a normal practical. While this practical is not assessed, it will help you gain the skills you need to successfully complete the remaining practicals that will count towards your grade. It is thus essential that you not only complete this, but ensure that you try to learn as much as possible in the process and take initiatives to find other sources of information when needed. \n",
    "\n",
    "If you feel that you need a little practice or review of Python, we recommend that you first read through the Python Guide and complete the four exercises it contains before starting Practical 0. While working through Practical 0, we encourage you to refer to the Python guide as needed to better understand different programming concepts, and also take the opportunity to speak with tutors to ask any questions you may have. All of the concepts introduced in this practical will be used to complete the six practicals and three practical assessments across the semester."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Part 1: Reviewing Python basics</SPAN>**\n",
    "\n",
    "### After completing Part 1 you should be familiar with:\n",
    "* How to perform basic operations on biological sequence data (i.e. DNA, RNA, proteins)\n",
    "\n",
    "### Programming lingo - let's start from the bottom and work our way up\n",
    "\n",
    "An **<span style=\"color:blue\">object</span>** is something that exists in a defined manner – a unique instance of a data structure. It is recognised by the computer and the computer knows all its features because they are described in a data structure. This concept is quite abstract but you need to understand that a computer only knows what you tell it and if you can’t describe an object – you can’t build code. Luckily, in Python, a lot of objects come predefined: integer, string, double, float, list, dictionary, set, tuple, etc. You should know what each of these objects is and how to use them. Each of these object types is actually a data structure and there are special operations you can do to use these objects. \n",
    "\n",
    "In most cases, you will want to store an object using a specific name so you can keep track of it and pass it around your code. When you assign an object a name, you create a **<span style=\"color:blue\">variable</span>**.\n",
    "\n",
    "`Variable1 = Object1`\n",
    "\n",
    "Now we have named our objects creating variables, what do we do to use the variables? We write a function. A **<span style=\"color:blue\">function</span>** is a list of instructions (code) that we have defined and given a name. You don’t have to organise your code into functions but good programmers always do. A function can take input in the form of variables, execute code, and then return an answer. A function is flexible and can also work in a number of different ways: input -> process -> output is just standard. Below is an example of a function.\n",
    "\n",
    "```python\n",
    "def revString(string):  \n",
    "    print(string)\n",
    "    nString = \"\" #create an empty string  \n",
    "    for c in string[::-1]: #iterate backward over the string  \n",
    "        nString += c #add each character to the new string  \n",
    "    print(nString)\n",
    "```\n",
    "\n",
    "Once we have loaded this function (selecting *Run* in Spyder), we can use it in the following way:\n",
    "\n",
    "```python    \n",
    ">>> revString('abcdefg')\n",
    "abcdefg\n",
    "gfedcba\n",
    "```\n",
    "\n",
    "You should now work through the following programming exercises, with the help of the tutor if necessary, to further practice Python programming. We are first going to implement basic operations on biological sequence data (specifically DNA, RNA and proteins).\n",
    "\n",
    "**<span style=\"color:darkred\">If at any point throughout these exercises you are not sure where to start, or you're already feeling a bit lost, we suggest that you review the Python Guide which covers all the basics mentioned above in more detail. introduction.py and bonus_code.py also contain a number of examples that you can run using Spyder.</span>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "   \n",
    "**Exercise 1 : Create a simple counting function** \n",
    "\n",
    "Create a string using the DNA sequence *exercise1_seq* below, which can also be found in *introduction.py*. Use the `string`’s [`count`] (https://www.tutorialspoint.com/python/string_count.htm) function to count the number of adenine (A), cytosine (C), guanine (G) and thymine (T) bases in the string. You might consider a `for` loop to iterate over each character that must be counted. Your function should print the counts out like so:\n",
    "\n",
    "```\n",
    "        Base counts:\n",
    "        Adenine: 1\n",
    "        Cytosine: 2\n",
    "        Guanine: 3\n",
    "        Thymine: 4\n",
    "```\n",
    "\n",
    "   *Hint: The count function works like this: `aString.count(\"countme\")`.*\n",
    "\n",
    "```\n",
    "AAAACCTCTCTGTTCAGCACTTCCTCTCTCTTGGTCTGGTCTCAACGGTCACCATGGCGAGACCCTTGGAGGAGGCCCTGGATGTAATAGTGTCCACCTTCCACAAATACTCAGGCAACGAGGGTGACAAGTTCAAGCTGAACAAGACAGAGCTCAAGGAGCTACTGACCAGGGAGCTGCCTAGCTTCCTGGGGAGAAGGACAGACGAAGCTGCATTCCA\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 2 : Store your function in a new module** \n",
    "\n",
    "Create a new *module* called *mysequence.py* (i.e. a new python file) and write a function called `baseCounts` that takes a string containing a DNA sequence input and does the same thing as in Exercise 1. Try it out on the above sequence.\n",
    "\n",
    "   *Hint: See the functions section of the Python reference guide for help on defining your own functions.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 3 : Expand the versatility of your baseCounts function** \n",
    "\n",
    "You will encounter a number of different alphabets (stored as strings in this exercise) when dealing with biological sequence. You do not want to restrict your function to only counting the four DNA bases. Modify your function so that in addition to a DNA sequence, it takes a string containing all of the allowed symbols in an alphabet. This string could be “ACGT” (DNA) or “ACGU” (RNA) or the 20 amino acids. It could also contain the alphabet in any order (i.e. “GCTA” is still valid for DNA). We will refer to this variable as *alphabet* (i.e. add a second argument to your function named alphabet). Your function should then be able to be used with any type of sequence depending on the alphabet. Again, try it out on the example sequence, experimenting with different alphabets to see the outcome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 4 : Modify the output format of baseCounts** \n",
    "\n",
    "Modify your `baseCounts` function so that instead of printing out the counts, it returns them in a dictionary. The keys should be the names of the different bases and the values the counts of each. Again, try it out on the example sequence.\n",
    "\n",
    "   *Hint: Remember that a dictionary is defined like this:*\n",
    "\n",
    "```\n",
    "aDict = {\"key1\": value1, \"key2\": value2}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Part 2: More complex Python structures</SPAN>**\n",
    "\n",
    "### After completing Part 2 you should be familiar with:\n",
    "* The structure of a class and how to access its functionality\n",
    "* The purposes and structures of constructors, class methods, and class variables\n",
    "* How to edit and use classes\n",
    "\n",
    "### Working with Classes\n",
    "\n",
    "Imagine you are working with sequence data, for example DNA. DNA could be represented as a string object but you want to store lots of sequences and you want to associate each sequence with specific features like an alphabet. You also want to be able to perform special operations specifically on your sequences. If you stick with the string data structure – you are limited in what you can do to your sequence. What you really want is a new data structure that will describe a sequence along with all of its special features – this is called a *class*.\n",
    "\n",
    "Open `guide.py`. What is in this file? \n",
    "\n",
    "Let’s review some of the terminology that is used in the guide.py module. The term `def` is used to describe a function or a class method. *A function is defined outside a class while a class method is defined within a class.* Using class methods is more complex than other functions as we must first create an object based on a class, called an instance. \n",
    "\n",
    "Indentation is key in Python and that is one way you can distinguish class methods and functions. The first line of a class method (containing the term `def`) will be indented underneath a line that starts with the term `class`, while the first line of a function (containing the term `def`) will not be indented. \n",
    "\n",
    "Another difference between functions and class methods is that class methods always use the term `self`. For example, you could have lots of `Sequence` instances, so `self` is an argument that tells the computer which `Sequence` instance to use. You **do not** specify what `self` is when you create a class instance or call a class method; it is built into the class and class methods. *The Python Guide introduces classes with additional detail and examples if you would like more information.*\n",
    "\n",
    "The `guide.py` module contains various classes, each with a different set of analysis tools/approaches for sequences and alignments. Consider the below questions:\n",
    "\n",
    "* How many *class methods* do each of the *seven classes* in the `guide.py` module have?\n",
    "\n",
    "* How many *functions* are in the `guide.py` module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 5 : Introducing the Alphabet Class**\t\n",
    "\n",
    "To work with the classes in the following exercises, you will need to import `guide.py`. If you don’t import the module you are interested in using, Python will not be able to recognise any of the functions or classes you are trying to use from the module. Run the following in the interpreter to import `guide.py`:\n",
    "\n",
    "```python\n",
    "from guide import *\n",
    "```\n",
    "\n",
    "This module contains a class for `Alphabet`. Following the `Alphabet` class, a number of standard alphabets are instantiated. List them. Can you access these alphabets from the interpreter?\n",
    "\n",
    "* What happens when you type DNA_Alphabet in the interpreter?\n",
    "* What happens when you type DNA_Alphabet.symbols?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 6 : Introducing the Sequence Class and its constructor**\t\n",
    "\n",
    "Use the code below to create a `Sequence` object (or instance) called `seq1` of the `Sequence` class.\n",
    "\n",
    "```python\n",
    ">>> seq1 = Sequence('AAAAAAAATTTT', DNA_Alphabet) \n",
    "```\n",
    "When you create an instance of a class, you run the code contained in a special type of class method called the **constructor** (`def __init__`) that is required for every class. \n",
    "\n",
    "* How many arguments (inputs) are required by the constructor?\n",
    "* How many arguments (inputs) can the constructor have in total?\n",
    "\n",
    "You can also override default arguments by specifiying the input for the arguments. Below is an example of how you could override the default input for the name argument in the instance `seq1` we just created\n",
    "\n",
    "```python\n",
    ">>> seq1 = Sequence('AAAAAAAATTTT', DNA_Alphabet, name = 'Sequence1') \n",
    "```\n",
    "What does this `seq1` instance actually look like? And how do we access the sequence information it contains? As shown in the case below, if you just type an instance in the interpreter then it will tell you that `seq1` is an object of the `Sequence` class, which in this case is stored at memory location `0x7fdff87b2220`. Now instead try printing `seq1` – how and why does the result change?\n",
    "\n",
    "```python\n",
    ">>> seq1\n",
    "<guide.Sequence at 0x7fdff87b2220>\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 7 : Working with class variables**\t\n",
    "\n",
    "A class object has class variables that are specific to each class object. Class variables can be accessed using `object.variable` to retrieve specific information stored in the instance. It is important to note that class variables can also be accessed by any class method or function within the class.\n",
    "\n",
    "Consider how the constructor of `Sequence` (i.e. `def __init__`) works. How many class variables does an instance of the `Sequence` class have? What happens if you just type one of its class variables, such as `alphabet`, in the interpreter and hit enter? Is `alphabet` a defined variable outside the `Sequence` class or does Python not recognise it?\n",
    "\n",
    "Now let's go through an example of how to access the class variable `alphabet` for the class object `seq2` below. In the first statement you will see that typing `seq1.alphabet` gives a similar output to what we saw in the previous exercise with the `Sequence` class object `seq1`. If you instead print `seq1.alphabet`, how does the result change? How and why does the information of `seq1.alphabet` and `seq2.alphabet` differ?\n",
    "\n",
    "```python\n",
    ">>> seq1.alphabet\n",
    "<guide.Alphabet object at 0x10230ff98>\n",
    ">>> print(seq1.alphabet)\n",
    "'ACGT'\n",
    "\n",
    ">>> seq2 = Sequence('AWAAAAAAGGVG', Protein_Alphabet)\n",
    "```\n",
    "\n",
    "Class variables are specific to each class instance and can be accessed as demonstrated above. What happens if you just type alphabet and hit enter? Is alphabet a defined variable outside the class or does Python not recognise it?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 8 : Using class methods** \n",
    "\n",
    "Now that we’ve covered class variables and constructors, let's look at class methods (the ones without `__` at the start - we'll look at those next). We have already noted the presence of `self` in all class methods. Class methods are specific to instances of the class itself. They have to be associated with an instance to obtain all required information (like class variables) and this association is achieved with the following notation – `Instance.class_method()`. \n",
    "\n",
    "Let's use the `Sequence.count()` class method as an example. Look at the `count` function within the Sequence class and how it accesses the `self.sequence` variable. Count the number of glycines (G) in `seq2` below.\n",
    "\n",
    "```python\n",
    ">>> seq2 = Sequence('AWAAAAAAGGVG', Protein_Alphabet)\n",
    ">>> count = seq2.count('G')\n",
    "```\n",
    "\n",
    "How does this `count` method compare to the one you designed in Exercises 1-4? Modify the `count` method in the Sequence class to output a *dictionary* recording the counts of all symbols in the alphabet like you did in Exercise 4.\n",
    "\n",
    "Once you have done that, create a new `Sequence` object from sequence *exercise8_seq*, which can also be found at the bottom of *introduction.py*. Then use `count` to print out the number occurrences of each amino acid as in Exercise 4. Make sure you understand the idea of classes and objects, because you will be using them a lot in the coming practicals.\n",
    "\n",
    "```\n",
    "MHSSIVLATVLFVAIASASKTRELCMKSLEHAKVGTSKEAKQDGIDLYKHMFEHYPAMKKYFKHRENYTPADVQKDPFFIKQGQNILLACHVLCATYDDRETFDAYVGELMARHERDHVKVPNDVWNHFWEHFIEFLGSKTTLDEPTKHAWQEIGKEFSHEISHHGRHSVRDHCMNSLEYIAIGDKEHQKQNGIDLYKHMFEHYPHMRKAFKGRENFTKEDVQKDAFFVNKDTRFCWPFVCCDSSYDDEPTFDYFVDALMDRHIKDDIHLPQEQWHEFWKLFAEYLNEKSHQHLTEAEKHAWSTIGEDFAHEADKHAKAEKDHHEGEHKEEHH\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 9 : Special class methods**\t\n",
    "\n",
    "When we write classes we can create special functions that will decide what a particular operator should do within a class. For example, we can change the normal function of the `str` operator within a class by defining it as a special class method using `def __str__`. All of these special functions will have names starting with “__”. Let's go through an example to explore these special class methods further. We will use some of the functions defined in `guide.py` to operate on lists of sequences, including reading sequences from, and writing sequences to, files. \n",
    "\n",
    "Create a list called `seqs` from the sequences in the file *mystery2.fa* using the function `readFastaFile()` from `guide.py`– note that it stores each fasta entry as a `Sequence` object within the list. While we know what `len(seqs)` does since `seqs` is a list, in the loop below we see that `len(seq)` is doing something different. This is because `seq` is not a list, but instead an instance of the `Sequence` class.\n",
    "\n",
    "```python\n",
    ">>> seqs = readFastaFile('mystery2.fa', Protein_Alphabet)\n",
    ">>> len(seqs)\n",
    "23\n",
    "\n",
    ">>> long_seqs = []\n",
    ">>> for seq in seqs:\n",
    "...     print(seq.name, len(seq), len(seq.alphabet))\n",
    "...     if len(seq) > 5000:\n",
    "...         long_seqs.append(seq)\n",
    "\n",
    "NP_078820.2 352 20\n",
    "NP_001171939.1 113 20\n",
    "NP_001171938.1 190 20\n",
    "...\n",
    ">>> writeFastaFile('mystery2_long.fa', long_seqs)\n",
    "```\n",
    "\n",
    "How was the function of the `len` operator changed within the `Sequence` class in `guide.py` using the special function `def __len__`?  What other special functions can you find in `Sequence`? List them, and provide a short example of how you can use them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 10 : Create a new function for the Sequence class** \n",
    "\n",
    "Write a `reverseComplement` function for the `Sequence` class that returns the reverse complement of a DNA sequence as a new `Sequence` object. (Remember that A &#8596; T and C &#8596; G.) Your function should check that the sequence is in fact a DNA sequence before proceeding, and print an *error* if it isn’t (taking the reverse complement of a protein doesn’t make biological sense).\n",
    "\n",
    "*Hint: It is possible to loop through the characters in a string in reverse. The following code does so while copying each character to a new `string` (in reverse). You could modify this so that it copies the complement of each base instead.*\n",
    "\n",
    "```python\n",
    "newSequence = ''\n",
    "for symbol in sequence[::-1]:\n",
    "    newSequence += symbol\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 11 : Extend your function to work with RNA** \n",
    "\n",
    "Modify your function so that it works for both DNA and RNA sequences.\n",
    "*Note: Adenine (A) and Uracil (U) are complements in RNA.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 12 : Importing modules** \n",
    "\n",
    "Make sure the module that contains the `Sequence` class imports `translate.py` (a different Python module). This way you will have access to all the variables in that module. They will prove helpful below. \n",
    "\n",
    "Provide an explanation of the difference between the following two alternative import statements to demonstrate your understanding of import statements.\n",
    "\n",
    "```python\n",
    "import translate\n",
    "```\n",
    "\n",
    "``` python\n",
    "from translate import *\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 13 : Write a more complex function** \n",
    "\n",
    "Write a new function for the `Sequence` class, called `translateDNA`, which will translate a DNA sequence into an amino acid sequence and allows for any of the six possible reading frames to be selected for translation.\n",
    "\n",
    "The `standardTranslation` variable in the `translate.py` file is a `dictionary` whose keys are DNA codons (i.e., sequences of three nucleotides) with values being the corresponding amino acid. Your function should use this dictionary to perform the translation. Like the reverse complement function, your translation function should print an error if the sequence is not DNA.\n",
    "\n",
    "Your function should accept two arguments in addition to the sequence:\n",
    "  * an integer 0, 1 or 2, indicating which reading frame to translate\n",
    "  * a `Boolean` specifying whether to translate in the forward or backward direction. \n",
    "\n",
    "You can use the function below as a framework to start from. This functions translates the DNA sequence starting from the 0 reading frame in the forward direction, and loops through the sequence three characters at a time using `while`. Note that the sequence length may not be a multiple of three, as shown below, and you might have to discard some incomplete codons. Next, you would need to modify your code to choose between all the forward reading frames. After that, implement reverse direction translation.\n",
    "\n",
    "*Hint: Print statements are extremely helpful to determine what your code is producing and how variables change as the code is executed.*\n",
    "\n",
    "```python\n",
    "seq = 'AGCTAGCT'\n",
    "i = 0\n",
    "while i < len(seq):\n",
    "    print('Codon', seq[i:i+3])\n",
    "    i = i + 3\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 14 : Test your new function** \n",
    "\n",
    "You have been given the mouse DNA sequence named *exercise14_seq*, which is also found at the bottom of *introduction.py*. It is strongly suspected that it contains a gene that codes for an important protein, but we don’t know where or what it is. See if you can find a possible protein coding sequence by looking at different frame shifts and the alternate strand. Save the most probable protein sequence. You may come back to it later...\n",
    "\n",
    "*Hint: The codon ATG is the most common \"Start\" message for translation. Most proteins are more than 100 amino acids long.* \n",
    "\n",
    "```\n",
    "GGCTAGAAGGATGGATGTTGTGAATGTTCCTGATGCAATGCAATCCTGTCACTAACATTAAAAACTTATCTTCCAACGCAAATCAAACGCCAAACAACCAACATAAAAGAACAAATGCTGCCTGCAAATCCACTCCCACATTGTCCGCTGTGGGTCCAGCTGCCGGGAAGGAGGCATAACCAGCTATCTTTGTGCTTCAGTGGGGCCTGGCTGGGTTTAGCTCAGGTGGTAACGATGCTCCCCCCGCGTGATATGGGATGAGAACTCGTACCTGTCCTGGCTGTGATAACCACACATGTTACACTCAAAGGGATCCCGAAAGCCATGGCAGCCATGGCAGCCCATGTGAATGGTATACATGACGTGATCCAGGAAGAGCACGCGGCAGTGTTCGCACTTGTACACCTTCAGCTGCTCGCCACTCGTGCTGACCACACGGAAGGCATCCTGCGAGTTCTCTGAGGCCGCCCTCAGCACCTCGTAGGCGCGCTGCTCCTCCTTGAGAGCCAGCCCATTGCGTGCATGCGGGTTGATGTGGTTGGTTAGGTAGATAAGGCCGCTGCGCTGTTCCTCCGCGTTGCTCTCTGTATCTGTGGAGTCTTGGCAGCTGTTGCTCGGGGAGGCCTCTCGCTCCGCTGACACAGACTTGGCCTTGGACAGCAGCAGCAAGTTATCCACGGCGTCCTGTGCTGAATGGTTGGACCGTGGGGGGCCATCTGAGGGGGGCTTGTGCAGCTGGTACATGGAGCTGATGACTGGCACCACCTCGGAGCTACCGGGGGGTGTCTGCACCAATGGGCGCAGGGACTCAGCCCCCAGGTAGTTGATGGCATTGTTGATGGCCTGGTCCATCACGTGGGATGTCATCATATCCTCCTTCTCATAGTTGGCACTGTCATAGGGCATGTCTGACAGGCACTTGTCTCCAAGAAATTTCTGAGGCATAGAGCTCTTACGTTTGGCGACATTGCTTGCCAGCCTGTCCAGGACAAGGGACCTCTCTGCTCCTATCTTGCACAGGTCTTCTGCCATCTCGTTGTGGTTAGTTTCTTCCTTAATGACTGGGCACACGCCCGGAAGGCCCATGCTTTCCAAGTAGTTGTGGCATCGCTCTTTATGCTCCTCTAAAGAGCTTCGCTGTTTATAGCTCCGGCCACAATATCCACATTTGTGAGGCTTACCAACGGAGTGCGTCCTCAGGTGGCCGGTGAGGGCGTCCCTCCGGCGGCAGGCATAGTTGCAAAGATGGCATTTGAAGGGCTTCTCACCCGAGTGCAGCTTGATGTGCCGCAGGAGGTTGCCTTTCTGGGTAAAGGAGGCCCCACACTGGTTGCACTGGAAAGGCCGTTCACCCATGCCTCGATCACTCTTGGAGTTCTGCTGTGCTCCAGAGGTAGTGGACAGGTCCTCAGGGACAGGCATGGGCTCATCCCCTTCATCTGGAGTGTCACTGACTGGGGGGCTCTCCTTTCCTGAAACTTGGGACATGTCTTGACCCTCATCGACATCCATTGTCTTCAGGTTATCAACAGACTTGTGCCTCCCCAAGG\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Part 3: Accessing online files and databases</SPAN>**\n",
    "\n",
    "### After completing Part 3 you should be familiar with:\n",
    "* How to perform specific searches of online biological sequence databases using Python\n",
    "* How to automate sequence download from online databases\n",
    "\t\n",
    "### Automating access to online resources\n",
    "\n",
    "One popular use of programming is to automate relatively trivial tasks that need to be performed repeatedly, e.g. reading long files of protein sequences, or looking up information for a large number of molecules over the internet. Some people refer to such programs as scripts because they are often written for a very specific purpose (that otherwise could be tedious), and not to be re-used.\n",
    "\n",
    "There are many biological sequence databases available online, some of which contain massive amounts of data. We will be using the NCBI (National Centre for Biotechnology Information), which indexes many nucleotide and protein sequence databases. We will first use the web site to search for a sequence and download it.\n",
    "\n",
    "### FASTA file format\n",
    "\n",
    "Biological sequences could just be supplied as the sequence by itself in a text file, but you can see why this would quickly get confusing when dealing with the massive amounts of sequence data that is available. Many file formats have been developed to store sequence data to help organisation. The most common format, and the one we will use is the FASTA format. A FASTA file can store one or more nucleotide or protein sequences, and looks like this:\n",
    "\n",
    "```\n",
    ">SequenceA\n",
    "AGCTCCGCGATATACCATAAAACGTA\n",
    ">SequenceB, with some more information\n",
    "ACGCTAGCTAGCTGCGCGCTATATATGCGCATAGATCGTCGCG\n",
    "AGTCGCTCGTAGCTAGTAGTCG\n",
    "```\n",
    "\n",
    "Each sequence in the file begins with a line starting with a > followed by the sequence’s name and possibly other unstructured information. The sequence itself starts on the following line, and ends when a line starts with a >, or the file ends.\n",
    "\n",
    "FASTA files are simple and easy to read. This also makes it easy to write code to read and write them. We have done this for you, and in *guide.py* you can use the `readFastaFile` and `writeFastaFile` functions to read and write lists of sequences from and to FASTA files. Note that `readFastaFile` always returns a list, even if there is only one sequence in the FASTA file. You will need to select a specific sequence from the list to use it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 14 : Create a new FASTA file of a database entry** \n",
    "\n",
    "Go to [http://www.ncbi.nlm.nih.gov/](http://www.ncbi.nlm.nih.gov/) and use the search form at the top to search for 'h1n1' (the swine influenza virus). The results page lists results from many different biological data resources, but we are interested in nucleotide sequences because we want DNA and RNA to test the functions you just wrote. Click on the nucleotide results on the left to get a list of nucleotide sequences. Pick one of the first few and click *FASTA* to get the sequence in the FASTA file format.\n",
    "\n",
    "Copy the FASTA data and then open a new file in a text editor (e.g. TextWrangler, TextEdit, SublimeText), paste in the data, and save the file as *h1n1.fasta* in the same directory as your other Python files.\n",
    "\n",
    "Be sure to note what is called the *GenBank accession number* above the FASTA data. Because there is so much biological data available, each sequence is identified by one or more unique identification numbers. Usually each database that stores a sequence refers to it by their own identification code. GenBank is a sequence database, and the number you see is the unique ID of this sequence in their database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 15 : Apply your functions to your new FASTA file** \n",
    "\n",
    "Using the `readFastaFile` function on the *h1n1.fasta* file you just saved, read in the sequence as you previously did in Part 2 of this practical. Try out the sequence functions you wrote in Exercises 10-14 on your new sequence. See if they return what you expect.\n",
    "\n",
    "```python\n",
    "seqs = readFastaFile('h1n1.fasta', DNA_Alphabet)\n",
    "seq = seqs[0] # Select first sequence in file\n",
    "```\n",
    "\n",
    "Do you remember why we supplied the filename first and then the variable DNA_Alphabet and where that came from? Where did that come from? If not, go back to Part 2 of this practical and/or review the *guide.py* file.\n",
    "\n",
    "*Note: if you saved several sequences in the file, you would access each by its index 0, 1, 2, … and so on.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 16 : Retrieve sequences directly with Python** \n",
    "\n",
    "The function `getSequence` in *guide.py* performs the retrieval of a single sequence by name. It needs a database parameter, which could be 'genbank' for searching in GenBank, or 'uniprot' for searching in Uniprot. The function will return an instance of the `Sequence` class if there is a match to the supplied query. \n",
    "\n",
    "Retrieve the DNA sequence identified by the ID `L03547` in GenBank and use your translation function to identify the most probable protein sequence. What is the protein sequence? Does it look familiar? It may be worth a visit to the GenBank web site…. Save it in FASTA format together with the protein sequence you found in Exercise 13. \n",
    "\n",
    "*Hint: Below is an example of how to retrieve a sequence.*\n",
    "\n",
    "```python\n",
    "seq = getSequence('L03547', 'genbank', DNA_Alphabet)\n",
    "print(seq.sequence)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 17 : Analyse the GenBank sequences** \n",
    "\n",
    "We are now going to analyse the composition of multiple sequences. Doing this by hand would quickly become tiresome. Use `getSequence` and then either by re-using the methods you developed earlier or by working out how to use the methods in `guide.py`, count the base composition of all the sequences you saved from your ‘h1n1’ search. Print out the overall frequency of each base as a percentage.\n",
    "\n",
    "*Hint: The `readFastaFile` function returns a list that you can loop through, and retrieve information about each sequence as you go. You will need to count up the total number of each base in the sequences, and (maybe not?) the total length of all the sequences, to calculate the percentages.*\n",
    "\n",
    "``` python\n",
    "# initialise counts here ...\n",
    "for sym in DNA_Alphabet:\n",
    "    for seq in seqs:\n",
    "        cnt += seq.count(sym)\n",
    "# process counts here ...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**Exercise 18 : Perform more complicated database searches** \n",
    "\n",
    "We suspect that so-called G-protein coupled receptors (GPCRs) have a higher proportion of hydrophobic amino acids, e.g. Leucine (L), Valine (V), Isoleucine (I) and Phenylalanine (F), being membrane associated proteins, compared to soluble proteins. Of course, looking at one example won’t convince a statistician; so to investigate this we will inspect all fruit-fly GPCRs proteins (i.e. proteins encoded by the fruit-fly genome, belonging to the GPCR family). Could you do this please?\n",
    "\n",
    "**You will need a few pointers:**\n",
    "\n",
    "* Uniprot can be searched for protein sequences like so:\n",
    "\n",
    "```python\n",
    "prot_ids = searchSequences('organism:7227+AND+family:G-protein+coupled+receptor', 'uniprot')\n",
    "```\n",
    "\n",
    "This should give you about 200 sequence identifiers as a list.\n",
    "\n",
    "Note that the first argument looks fairly complicated. Indeed, this follows a search language that Uniprot has made up for advanced searching. If you paste in the following into your web browser (one line):\n",
    "\n",
    "[http://www.uniprot.org/uniprot/?query=organism%3A7227+AND+family%3A\"G-protein+coupled+receptor\"](http://www.uniprot.org/uniprot/?query=organism%3A7227+AND+family%3A\"G-protein+coupled+receptor\") \n",
    "\n",
    "you will get the same entries. You can fine-tune your query 'string' by using Uniprot’s Advanced Search function and then copy it across to your Python code, replacing the percentage gunk with proper characters.\n",
    "\n",
    "* Find the amino acid sequence for each identifier and determine the percent of hydrophobic amino acids\n",
    "\n",
    "``` python\n",
    "count_hydrophobic = 0\n",
    "count_all = 0\n",
    "for id in prot_ids:\n",
    "    seq = getSequence(id, 'uniprot', Protein_wX)\n",
    "    count_all = count_all + len(seq)\n",
    "    count_hydrophobic = ...\n",
    "```\n",
    "\n",
    "* With that percentage in hand, you need to find what proportion holds for fruit-fly proteins in general ('organism:7227') but this could take a while if you look at all of the 37,000 or so proteins. Instead I suggest you look at every 200th protein. Here’s a trick:\n",
    "\n",
    "``` python\n",
    "for i in range(0, len(prot_ids), 200):\n",
    "    seq = getSequence(prot_ids[i], 'uniprot', Protein_wX)\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
