{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**<center style=\"color:#5F018B;font-size:24pt;weight:bold\">SCIE2100 Practical 1 (Week 3):</center>**\n",
    "\n",
    "**<center style=\"color:#5F018B;font-size:20pt;weight:bold\">Sequences, alphabets, databases, and homology</center>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt\">Aim</SPAN>**\n",
    "\n",
    "**After reading and completing this practical, you should:**\n",
    "* Be familiar with some of the Python modules that will be used for sequence analysis from now on, in this course\n",
    "* Know a few more things about classes in Python \n",
    "* Know how to access data both in the form of files and web resources and be aware of some of the issues involved\n",
    "* Have some experience in writing ‘scripts’\n",
    "\n",
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt\">Outline</SPAN>**\n",
    "\n",
    "[**<span style=\"color:black\">Part 1: More work with classes</span>**](#part1)\n",
    "* Reviewing the FASTA file format\n",
    "* An extended `Sequence` class and its friends\n",
    "* Exercises 1-2\n",
    "\n",
    "[**<span style=\"color:black\">Part 2: Sequence databases</span>**](#part2)\n",
    "* Online databases\n",
    "* Retrieving database entries with Python\n",
    "* Exercises 3-5\n",
    "\n",
    "[**<span style=\"color:black\">Part 2: Sequence alignments</span>**](#part3)\n",
    "* Multiple sequence alignments\n",
    "* Exercise 6\n",
    "\n",
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Introduction</SPAN>**\n",
    "\n",
    "### Answer submission\n",
    "\n",
    "All your answers should be submitted through [Coder Quiz](http://bioinf.scmb.uq.edu.au:81/coderquiz/) before **<span style=\"color:blue\"> Friday 19th of March, 2021 at 12PM. </span>**\n",
    "\n",
    "To receive full marks for this practical, you must answer ALL questions. Coder Quiz has the ability to check correctness of provided answers (where applicable) prior to submission. This means that as you progress through the practical, you can use Coder Quiz to determine whether your answer is correct or not. Coder Quiz will also warn you if your submission contains incorrect answer. You can submit incorrect answers and still receive full marks if the answer shows a reasonable attempt at answering the question.\n",
    "\n",
    "If you leave the answer to a question blank, your submission will be marked incomplete and receive 0 marks. If you cannot answer a question, it is sufficient to provide an explanation of why the question confused you, the approach you used to attempt to solve the question, or how you think the question should have been answered. Demonstrating you attempted the question will gain full marks whereas a blank field will receive no marks. Coder Quiz will warn you prior to submission if you have left blank fields and will receive 0 marks. Simply enter any explanation into the field relating to the question you haven't completed.\n",
    "\n",
    "Coder Quiz may ask you to provide code for an answer, in this case you just need to save the relevant section of code as a Python file and then upload this file to Coder Quiz. It is for visual inspection of your attempt and does not need to run (i.e. you do not need to include import statements) but should be commented and understandable by a tutor.\n",
    "\n",
    "You may submit as many attempts to Coder Quiz as you like, the most recent submission before the due date will be the attempt that is assessed. On the 'View Submissions' page you may view your latest submission or all of your attempts to ensure that you have answered every question.\n",
    "\n",
    "Coder Quiz does not save or retrieve partial attempts, so we recommend storing your work and answers in a seperate file and using Coder Quiz to validate and submit once you are complete.\n",
    "\n",
    "Coder Quiz submissions will be validated by a tutor prior to release of marks."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "<a id=\"part1\"></a>\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Exercises part 1 : More work with classes</SPAN>**\n",
    "\n",
    "### Reviewing the FASTA file format\n",
    "\n",
    "In *Practical 0* we learned that the most common format for permanently storing sequence data, and the one we will use predominately is the FASTA format. A FASTA file can store one or more nucleotide or protein sequences, and looks like:\n",
    "\n",
    "```\n",
    ">SequenceA\n",
    "AGCTCCGCGATATACCATAAAACGTA\n",
    ">SequenceB, with some more information\n",
    "ACGCTAGCTAGCTGCGCGCTATATATGCGCATAGATCGTCGCG\n",
    "AGTCGCTCGTAGCTAGTAGTCG\n",
    "```\n",
    "\n",
    "Each sequence in the file begins with a line starting with a `>` followed by the sequence’s name and possibly other unstructured information. The sequence itself starts on the following line, and ends when a line starts with a `>`, or the file ends. \n",
    "\n",
    "### An extended `Sequence` class and its friends\n",
    "\n",
    "The module `sequence.py` defines a few important classes and imports a few from other modules. Most of the classes are defined in more limited forms in `guide.py` as well, including the `Sequence` class that we previously worked with in Practical 0. In the `sequence.py` module we now have a new `Sequence` class that has additional functionality compared to what we previously experienced. For example, the new `Sequence` class does not require that you specify the alphabet (e.g. `RNA_Alphabet`) when creating a class object as it will default to taking a guess instead. \n",
    "\n",
    "Fire up a Python interpreter (e.g. IDLE or Spyder) and import `sequence.py` to test it out for yourself with the examples below:\n",
    "\n",
    "``` python\n",
    ">>> from sequence import *\n",
    ">>> seq1 = Sequence('AAAAAAAAGGGG')\n",
    ">>> print(seq1.alphabet)\n",
    "('A', 'C', 'G', 'T')\n",
    ">>> seq2 = Sequence('AAAAAAAAGGUG')\n",
    ">>> print(seq2.alphabet)\n",
    "('A', 'C', 'G', 'U')\n",
    ">>> seq3 = Sequence('AWAAAAAAGGVG')\n",
    ">>> print(seq3.alphabet)\n",
    "('A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y')\n",
    "```\n",
    "\n",
    "However, you can of course override the 'guess'. In the above examples, `seq1` was guessed to be a DNA sequence although based on its sequence it could also be an RNA sequence. Python allows parameters to be set 'by default' and be overridden (you would have seen how this is done for `__init__`). Let's override the `Sequence` class's guess.\n",
    "\n",
    "```python\n",
    ">>> seq4 = Sequence('AAAAAAAAGGGG', RNA_Alphabet)\n",
    ">>> print(seq4.alphabet)\n",
    "('A', 'C', 'G', 'U')\n",
    "```\n",
    "\n",
    "Of course, once you start taking responsibility for what the alphabet is, it may lead to an error if the alphabet does not capture what symbols the sequence actually contains. Let's override the default guess by introducing a new alphabet not found in `sym.py`. What happens if new alphabet does not reflect the sequence?\n",
    "\n",
    "```python\n",
    ">>> seq5 = Sequence('AAAAAAAAGGGG', Alphabet('AGU'))\n",
    "('A', 'G', 'U')\n",
    ">>> seq6 = Sequence('AAAAAAAAGGGG', Alphabet('ATU'))\n",
    "Traceback (most recent call last):\n",
    "...\n",
    "RuntimeError: Invalid symbol: G in sequence\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**Exercise 1.** \n",
    "\n",
    "When you import the module `sequence.py`, it automatically imports the `sym.py` module as well. In `sym.py`, a class for `Alphabet` is defined within which some standard alphabets are instantiated.\n",
    "\n",
    "**<span style=\"color:blue\">Submit: List the standard alphabets defined in sym.py (e.g. Bool_Alphabet). Please ensure your spelling is correct and your answers are separated by a comma (,).</span>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**Exercise 2.** \n",
    "\n",
    "A number of methods are defined in `sequence.py` that operate on lists of sequences, including `readFastaFile` that we used in Practical 0. You should have a FASTA file `mystery1.fa` in your directory. Study the below example that uses the FASTA file `mystery1.fa` in your directory. \n",
    "\n",
    "We know what `len(seqs)` does, because `seqs` is a `list`. But what does `len(seq)` do? `seq` is not a `list`, or a `set`… it is a `Sequence`. In fact, when we write classes we can decide what the `len` operator should do, by defining the special `def __len__` method. Look it up. \n",
    " \n",
    "```python\n",
    ">>> seqs = readFastaFile('mystery1.fa')\n",
    ">>> len(seqs)\n",
    "27\n",
    ">>> long_seqs = []\n",
    ">>> for seq in seqs:\n",
    "...    print(seq.name, len(seq), len(seq.alphabet))\n",
    "...    if len(seq) > 250:\n",
    "...        long_seqs.append(seq)\n",
    "\n",
    "Q9H9L7 192 20\n",
    "P41223 144 20\n",
    "Q13352 177 20\n",
    "...\n",
    ">>> writeFastaFile('mystery1_long.fa', long_seqs)\n",
    "```\n",
    "\n",
    "**<span style=\"color:blue\">Submit: List the 'special' methods from the sequence class in the first field. Just enter the method name for each, for example, for `__len__(self)` you would enter `__len__`.  In the second, provide an example of each function's use using text, not code.</span>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "<a id=\"part2\"></a>\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Exercises part 2 : Sequence databases</SPAN>**\n",
    "\n",
    "### Online databases\n",
    "\n",
    "There are many biological sequence databases available online, some of which contain massive amounts of data (and you would not want to store all of that data on your own hard disk).\n",
    "\n",
    "NCBI [http://www.ncbi.nlm.nih.gov](http://www.ncbi.nlm.nih.gov) is one authoritative source for nucleotide sequence data and we referred to this in *Practical 0*. NCBI also stores data for proteins but the richest source of protein data is available from Uniprot ([uniprot.org](uniprot.org)). Uniprot naturally links with complementary sources of information, including the so-called Gene Ontology.\n",
    "\n",
    "### Retrieving database entries with Python\n",
    "\n",
    "We will often find ourselves wanting to search for proteins (and their sequences) in Uniprot, and the exercise below will help us develop strategies for doing that in an automated manner with Python, without a web browser.\n",
    "\n",
    "Let's first go to the Uniprot web site, and find 'RNS1_ARATH' Ribonuclease 1 in Arabidopsis thaliana. The information is broken down into sections. Study the information and focus on its sequence annotation. Uniprot says that it contains a so-called Signal peptide at positions 1-22. \n",
    "\n",
    "Now, let's try retrieving the sequence of this single gene, given its identifier (e.g. RNS1_ARATH), from Uniprot using `sequence.py`:\n",
    "\n",
    "```python\n",
    ">>> rns1 = getSequence('RNS1_ARATH', 'uniprot')\n",
    ">>> print(rns1)\n",
    "```\n",
    "\n",
    "To better understand what is contained in the rns1 variable, can you use code to calculate how many Serines (S) it has?\n",
    "\n",
    "*Hint: Read `sequence.py`, find `getSequence`, understand what it returns and what methods you can apply to the `rns1` object.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**Exercise 3 : Exploring database identifiers** \n",
    "\n",
    "Create a new Python file, say `mystery.py`. Import `sequence.py`. Using the sequence module, write code that loads the FASTA file called `mystery2.fa` which contains protein sequences from online databases. Online databases have different identifiers for their respective sequences. For example, in Ensembl, a human gene may have an identifier like ENSG00000162599. Iterate through the sequences in `mystery2.fa`, and print out their names.\n",
    "\n",
    "**<span style=\"color:blue\">Submit: How many different types of identifiers are associated with the sequences in mystery2.fa? (Submit the first two letters common to the identifiers). Which databases do these identifiers map to?</span>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**Exercise 4 : Retrieving many sequences online and with Python** \n",
    "\n",
    "Now let's compare two different approaches to downloading a large number of sequences. You will first download the sequences online using a webtool, and then you will use Python to programmatically download the sequences.\n",
    "\n",
    "Let's first retrieve the sequences online using a webtool. Visit the [Uniprot](http://www.uniprot.org/) website online and use 'Advanced Search' techniques to find all proteins belonging to *Arabidopsis thaliana* that contain a signal peptide. For the purpose of completing this practical without taking too long, restrict the length of the protein to 700 amino acids or more. Download your matches from Uniprot as a FASTA file called `sigpep_at.fa`. \n",
    "\n",
    "*Hint: Advanced searches were introduced in Practical 0.*\n",
    "\n",
    "Now, let's write a Python code to programmatically perform another search for *Arabidopsis thaliana* proteins in the Uniprot database, but this time searching for proteins involved in 'Lipid metabolism' (keyword search). If you're unsure of the search syntax, an example of the signal peptide advanced search you did using the webtool is below.\n",
    "\n",
    "``` python\n",
    ">>> spat = searchSequences('\"signal+peptide\"+AND+organism:3702+AND+length:[700+TO+*]')\n",
    "```\n",
    "\n",
    "Cryptic? This string is the result of the advanced search on Uniprot. It forms part of the URL used by Uniprot search and should also appear in the search bar of the Uniprot webpage. So, play with Uniprot search and note the search string in your web browser. Except for replacing all spaces with ‘+’, you can ‘cut-and-paste’ that string into your Python code.\n",
    "\n",
    "The function `searchSequences()` returns a list of identifiers which meet our search criteria and **not** the actual sequence information. We can loop over the result of `searchSequences` and use another function, `getSequence()`, to create a new list of Sequence objects (which will contain the actual sequence) from our original list of identifiers. Save the results from your search in a FASTA file called `lipmet_at.fa` using the `writeFastaFile` function demonstrated in Exercises part 1.\n",
    "\n",
    "As getting sequence data using the `getSequence()` function in `sequence.py` can take a while when you have many sequences, you might want to make sure the code works before downloading the entire set of sequences. For example, you could limit your initial search to larger proteins (> 700 amino acids) or first try it out on a subset of ten sequences from your list.\n",
    "\n",
    "\n",
    "**<span style=\"color:blue\">Submit: How many entries are in each of your fasta files (sigpep_at.fa and lipmet_at.fa)? Submit the code you used to identify these sequences with comments describing the process.</span>**\n",
    "\n",
    "*Hint: You do not actually have to download the sequences to know how many would be in the fasta files. (e.g. the number of IDs in spat will equal the length of sigpep_at.fasta)*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**Exercise 5.** \n",
    "\n",
    "Triacylglycerols (TAGs) are an important reserve of carbon and energy in Eukaryotes. Triacylglycerol (TAG) lipases have been thoroughly characterized in mammals and microorganisms. By contrast, very little is known about plant TAG lipases. \n",
    "\n",
    "We expect TAGs to have both a **signal peptide** and be involved in **lipid metabolism**. Hence, we would expect that any Triacylglycerol lipase(s) should be in both of the FASTA files retrieved in *Exercise 4*. Let's check if there any TAG lipases in *A. thaliana* and, if so, how many there are.\n",
    "\n",
    "One approach for answering this would be to store their IDs in **sets** by extracting them from the FASTA files. Then, you could find the IDs that are found in both sets using an intersection method for sets of strings. Sets are a python data structure you should familiarise yourself with! Set theory is useful in programming as it can allow easy comparison of data. See below for an example:\n",
    "\n",
    "```python\n",
    ">>> seqs1=readFastaFile('sigpep_at.fa')\n",
    ">>> seqs2=readFastaFile('lipmet_at.fa')\n",
    ">>> ids1=[];ids2=[]\n",
    ">>> ... ADD 1-2 LINES OF CODE HERE: to add the ids from sigpep to ids1 \n",
    ">>> ... ADD ANOTHER 1-2 LINES OF CODE HERE: to add the ids from lipmet to ids2\n",
    ">>> ids1=set(ids1)\n",
    ">>> ids2=set(ids2)\n",
    ">>> common_ids=list(ids1.intersection(ids2)) #the clever method\n",
    ">>> print(common_ids)\n",
    ">>> for seq in seqs1:\n",
    ">>> ... ADD 1-2 LINES OF CODE HERE: to print the details of identical IDs \n",
    "```\n",
    "\n",
    "**<span style=\"color:blue\">Submit: How many TAG lipases did you find? Submit your code with comments describing the process.<span style=\"color:blue\">**\n",
    "    \n",
    "*Hint: If your internet connection was too slow to download the sequences above, then instead of using the fasta files, you can just directly use the IDs outputted from searchSequences().*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "<a id=\"part3\"></a>\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Exercises part 3 : Sequence alignments</SPAN>**\n",
    "\n",
    "\n",
    "### Multiple sequence alignments\n",
    "\n",
    "Aligning sequences to one another lets us see where sequences are similar or different. Similarity in sequences may indicate that the sequences have the same structure, perform the same function, or that they have evolved from the same ancestor.\n",
    "\n",
    "`Alignment` class stores an alignment of sequences. To get a feel for the data, load one of the `p450.aln` or `gpcr.aln` files, and print the alignment.\n",
    "\n",
    "\n",
    "```python\n",
    ">>> from sequence import *\n",
    ">>> aln = readClustalFile('p450.aln', Protein_Alphabet)\n",
    ">>> print(aln)\n",
    "Q27499  AGMETTSNTLNWALLYVLRNPEVRQKVYEELDATINESQRLANLVPFSIGKRQCCPGEGLAKMELL\n",
    "Q16872  AGTETTSTTLRYALLLLLKHPEVTAKVQEEIEAVVHEVQRYIDLMPFSAGKRICCVGEALAGMELF\n",
    "...\n",
    "```\n",
    "\n",
    "Glancing over the alignment you will observe that each sequence occupies its own row, and that columns tend to contain the same or similar amino acids — a result of evolutionary conservation. To facilitate the inspection of this data, it is common to colour amino acids according to their physico-chemical properties. There’s a method for writing the alignment to an HTML file:\n",
    "\n",
    "```python\n",
    ">>> aln.writeHTML('p450.html')\n",
    "```\n",
    "\n",
    "You can view this file in a web browser. (Open file; or simply double-click in the file explorer.) Here’s a snippet:\n",
    "![p450 alignment snippet](p450.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**Exercise 6 : Investigating an alignment of GPCRs** \n",
    "\n",
    "Load `gpcr.aln` and write it to a HTML file to view as was shown above. The file contains the full sequences of 23 so-called G protein coupled receptors (GPCRs). All GPCRs contain seven transmembrane domains (i.e. sections of 15-25 consecutive residues that embed in lipid membrane). \n",
    "\n",
    "Investigate and describe the meanings of the colours used in the alignments by default. (`writeHTML` uses a colour 'annotation' that is defined for `Protein_Alphabet` in `sym.py`, halfway down the file.) Find out which amino acids are considered hydrophobic and then colour all of them blue by changing the colour annotation in `sym.py`. Colour the other residues white. \n",
    "\n",
    "In the following diagram of what a GPCR protein looks like we can see that it bends and crosses over the lipid membrane seven times (as indicated by the numbers in the diagram).\n",
    "\n",
    "<img src=\"gpcr.png\" alt=\"GPCR diagram\" style=\"width: 400px;\"/>\n",
    "\n",
    "If you're not familiar with the hydrophobic properties of a lipid membrane, look them up. Based on this knowledge we know that if we took the sequence of amino acids that make up this protein, then even without knowing exactly what the amino acids are, we know that anytime it crosses through the membrane it will contain mostly hydrophobic amino acids.\n",
    "\n",
    "On your alignment, identify the boundaries of the transmembrane domains.\n",
    "\n",
    "\n",
    "**<span style=\"color:blue\">Submit: Describe the physico-chemical properties represented by each default colour used in the alignment.</span>**\n",
    "\n",
    "**<span style=\"color:blue\">Submit: Show your own 'hydrophobic' colour scheme (as a list of affected amino acids) along with a screen capture of your GPCR alignment.</span>**\n",
    "\n",
    "**<span style=\"color:blue\">Submit: Provide the rough boundaries of the fifth transmembrane domain.</span>**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "**<SPAN STYLE=\"color:#5F018B;font-size:18pt;weight:bold\">Assessment</SPAN>**\n",
    "\n",
    "### Be sure to provide your responses to the items marked with Submit through Coder Quiz by the due date !\n",
    "\n",
    "------------------------------------------------------------------------------------------------------------------------------\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
